### ISSUE TITLE
[Question / Discussion] PowerCollections in the Full .NET Framework

### ISSUE THREAD

**Issue by airbreather** [NetTopologySuite/NetTopologySuite#73]
In every case but `BigList<T>`, the full .NET Framework provides a direct built-in replacement for PowerCollections types used in NTS.  Some of them ([SortedDictionary&lt;TKey, TValue&gt;](https://msdn.microsoft.com/en-us/library/f7fta44c.aspx)) only require .NET 2.0+.  Others ([HashSet&lt;T&gt;](https://msdn.microsoft.com/en-us/library/bb359438.aspx)) require .NET 3.5+.

In one case, replacing the implementation of a method from PowerCollections's `OrderedDictionary<TKey, TValue>` with [something](https://gist.github.com/airbreather/8b8544f9c4babcc490ba) based on the [implementation](https://github.com/dotnet/corefx/blob/cd235dbf9d06d64b06643007de13307bba5b92be/src/System.Collections/src/System/Collections/Generic/SortedSet.cs#L209-L244) from the built-in `SortedDictionary<TKey, TValue>` took a significant chunk off the time from an operation.  Specifically, it eliminated between 13%-14% of the total time spent in `Geometry.Union(Geometry)` from `CascadedPolygonUnion`.  I did not go in-depth to see if there were other places where we get penalized for using PowerCollections instead of a built-in.

I wonder if it would be appropriate to revisit whether or not to include PowerCollections as a dependency in targets where there's a built-in equivalent for everything?  Built-in types tend to be highly optimized.  Specifically:

1. For [QuadEdgeSubdivision](https://github.com/NetTopologySuite/NetTopologySuite/blob/0f56e478f91a0a40e4b75203ade55908990fc4ce/NetTopologySuite/Triangulate/QuadEdge/QuadEdgeSubdivision.cs), replace `ISet` with `HashSetQ` and change the `#if NET40` to `#if NET35` (`ISet` shows up once, in a private method signature that's guaranteed to always have a `HashSetQ`, so this is not a breaking change).
2. For anything that uses `OrderedDictionary<TKey, TValue>`, use `SortedDictionary<TKey, TValue>` instead, `#if NET20`.
3. For anything that uses `OrderedSet<T>`, use `SortedSet<T>` instead, `#if NET20`.
4. For anything that uses `Set<T>`, use `HashSet<T>` instead, `#if NET35`.
5. For anything that uses `BigList<T>`, re-evaluate whether or not it's actually better than `List<T>`.
    1. For the [usage](https://github.com/NetTopologySuite/NetTopologySuite/blob/0f56e478f91a0a40e4b75203ade55908990fc4ce/NetTopologySuite/Index/Strtree/AbstractSTRtree.cs#L111-L114) in `AbstractSTRtree`, I think it's almost certainly a mistake to pick this over `List<T>`.  The `List<T>` constructor that would be used here is optimized for `ICollection<T>`-typed arguments (which this is), so initializing it would not require the array copies that `BigList<T>` exists to save us from.
    2. This leaves just the [usage](https://github.com/NetTopologySuite/NetTopologySuite/blob/0f56e478f91a0a40e4b75203ade55908990fc4ce/NetTopologySuite/Planargraph/DirectedEdgeStar.cs#L16) in the `PlanarGraph` version of `DirectedEdgeStar`.
        1. Do you have measurements that show that using `BigList<T>` here tends to result in favorable performance characteristics, as compared to `List<T>`?
        2. If so, can you please point me to them and a list of steps that can be used to reproduce?
        3. A polygonization for a [much more complicated line string](https://gist.github.com/airbreather/f3a63abc8464fda34d6b) than what I could find in the test repositories is still blazingly fast.  Running it 5000 times with a dramatically increased sampling rate in the profiler from what I usually like to use indicates **no** visible difference between `List<T>` and `BigList<T>` when I change it.  I expect that I'm not using something complicated enough to see the difference, though.

I did not look at the tests.

To summarize: on targets that have both `NET20` and `NET35` defined, PowerCollections is redundant and sometimes significantly slower.  What are your thoughts about just conditionally replacing the implementations with the built-ins and removing the reference?

**Comment by DGuidi** [comment]
> What are your thoughts about just conditionally replacing the implementations with the built-ins and removing the reference?

I'm unable to retrieve some refs now, but some tim ago I've tried to replace PowerCollections with .NET BCL stuff and I seen that some tests are failing, due to a different internal behavior...

**Comment by DGuidi** [comment]
ok found some refs:
1. [PowerCollections thread on GoogleGroups](https://groups.google.com/d/msg/nettopologysuite/rG5yMawWUQ4/qXDIIls2oesJ)
2. [Google Code Issue 154](https://code.google.com/p/nettopologysuite/issues/detail?id=154)

**Comment by airbreather** [comment]
@DGuidi: I see.  Do you have more information about what the "sorting issue" is with LineSequencer?

Running all the tests in classes named `LineSequencerTest` all pass after applying the obvious change.  I don't exactly know what I need to do to run the test in `LineSequencerFileTest`, or I would have run that too.

**Comment by airbreather** [comment]
Ignoring what's needed to actually remove the reference to PowerCollections on targets that have both NET20 and NET35 defined, this is what I mean when I say the "obvious changes": airbreather/NetTopologySuite@f1fe570fee0874bdd29b7569c62ba7825285809f.

**Comment by DGuidi** [comment]
> Running all the tests in classes named LineSequencerTest all pass after applying the obvious change.

Tried also with xml tests with console runner?
I'll try to check your fork as soon as possible.

**Comment by DGuidi** [comment]
all tests looks ok, actually.

**Comment by DGuidi** [comment]
> this is what I mean when I say the "obvious changes"

sorry, maybe I'm a but confused, but your code uses standard BCL objects only for 2.0 code.
and SortedSet<T> it's [4.0+ only](https://msdn.microsoft.com/library/dd412070(v=vs.100).aspx) so your code won't actually compile.
Your LineSequencer code isn't modified, and I'm unable to made the changes you suggests.

**Comment by airbreather** [comment]
Oh, crap, you're right.  `SortedSet<T>` is in fact 4.0+.  Looking into it again.

**Comment by airbreather** [comment]
OK, airbreather/NetTopologySuite@02bb1e6 *should* fix that, but as before, I can't be 100% sure because as before, I clearly haven't actually compiled with .NET 2.0.

I think there should probably be some build configurations in the solution and projects for all the different supported platforms, so that I don't have to either guess, screw with the .csproj files, create a "dummy" PR just to get Travis to test it, or install my own TeamCity to use that .proj / .targets stuff.  I guess that's another issue, though.

**Comment by airbreather** [comment]
Running tests on other platforms was harder than it sounds.  Anyway, I get what you're talking about now.

On .NET 3.5, `NetTopologySuite.Tests.NUnit.Operation.LineMerge.LineSequencerTest.Test2SimpleLoops` is the only test that fails.  Reverting just the `BigList<T>` change in `DirectedEdgeStar` makes it pass.

.NET 3.5 `List<T>`:

    MULTILINESTRING ((0 10, 0 0), (0 0, 0 10), (0 0, 0 20), (0 20, 0 0))

Expected, .NET 4.0 `List<T>`, `BigList<T>`:

    MULTILINESTRING ((0 10, 0 0), (0 0, 0 20), (0 20, 0 0), (0 0, 0 10))

Doing some more digging, it turns out that the problem is in the `SortEdges` method.  Java's `ArrayList` (which JTS [uses here](https://github.com/topobyte/jts/blob/7bc98b47623b6dc388eb0f9f0cd306cb4e9ab0c5/jts/java/src/com/vividsolutions/jts/planargraph/DirectedEdgeStar.java#L53)) sorts using a guaranteed stable sort, whereas .NET's `List<T>` does not guarantee that the sort will be stable.  My guess is that `BigList<T>` also guarantees a stable sort.  The BCL team [occasionally changes](http://stackoverflow.com/a/12461920/1083771) the implementation of `Array.Sort` for various reasons, and to my knowledge it has **never** guaranteed it would be stable.

If this is the only issue, then there are really two options:

1. Implement a stable sort so the **two** places that use `BigList<T>` can be replaced by `List<T>`, or
2. Keep PowerCollections around because it gives us a stable sort that exactly those two places rely on.

It really seems to me like we might as well just add the ability to stable-sort a list, which I'm working on right now.  It won't be as bad to look at as that probably makes it sound.

**Comment by DGuidi** [comment]
Removing powercollections should be the preferable way, at least to avoid licensing issues, so thanks for your investigations!

----- Messaggio originale -----
Da: "airbreather" <notifications@github.com>
Inviato: ‎23/‎01/‎2016 20:25
A: "NetTopologySuite/NetTopologySuite" <NetTopologySuite@noreply.github.com>
Cc: "Diego Guidi" <diegoguidi@gmail.com>
Oggetto: Re: [NetTopologySuite] [Question / Discussion] PowerCollections inthe Full .NET Framework (#73)

Running tests on other platforms was harder than it sounds. Anyway, I get what you're talking about now.
On .NET 3.5, NetTopologySuite.Tests.NUnit.Operation.LineMerge.LineSequencerTest.Test2SimpleLoops is the only test that fails. Reverting just the BigList<T> change in DirectedEdgeStar makes it pass.
.NET 3.5 List<T>:
MULTILINESTRING ((0 10, 0 0), (0 0, 0 10), (0 0, 0 20), (0 20, 0 0))
Expected, .NET 4.0 List<T>, BigList<T>:
MULTILINESTRING ((0 10, 0 0), (0 0, 0 20), (0 20, 0 0), (0 0, 0 10))
Doing some more digging, it turns out that the problem is in the SortEdges method. Java's ArrayList (which JTS uses here) sorts using a guaranteed stable sort, whereas .NET's List<T> does not guarantee that the sort will be stable. My guess is that BigList<T> also guarantees a stable sort. The BCL team occasionally changes the implementation of Array.Sort for various reasons, and to my knowledge it has never guaranteed it would be stable.
If this is the only issue, then there are really two options:
Implement a stable sort so the two places that use BigList<T> can be replaced by List<T>, or 
Keep PowerCollections around because it gives us a stable sort that exactly those two places rely on. 
It really seems to me like we might as well just add the ability to stable-sort a list, which I'm working on right now. It won't be as bad to look at as that probably makes it sound.

**Comment by airbreather** [comment]
Copied from the description of #75:

For the future, here are the things keeping .NET 2.0 and PCL from dropping it entirely:

We need a hash-based set type. It's possible to emulate HashSet<T> by wrapping a Dictionary<T, bool>. I didn't want to do this right away.
We need a way to do a stable-sort. One way to do it would be to add a utility method that wraps each element of an IEnumerable<T> in an ItemWithIndex<T> whose comparison implementation uses the original T's comparison, and then uses the index from the original sequence as a tiebreaker, then sort it and select the T elements of the sorted set back out.
Of course, another way would be to just go ahead and implement a stable sort within NTS, which is not very appealing either.
We need an ordered dictionary type. As mentioned above, it's not actually all that easy to do this. I have some ideas, but certainly none of them are worth me holding up this PR for. Even if it looks utterly obvious that EdgeList, for example, should be OK with using a hash-based dictionary, it's not. Because OrientedCoordinateArray does not have Equals / GetHashCode overrides that are consistent with its comparison behavior.
